server/src/test/java/org/elasticsearch/cluster/routing/ShardRoutingTests.java

```diff
@@ -11,16 +11,178 @@
+            state != ShardRoutingState.UNASSIGNED && randomBoolean(),
+            state
+        );
+    }
+
+    @Override
+    protected ShardRouting mutateInstance(ShardRouting instance) throws IOException {
+        return switch (randomInt(3)) {
+            case 0 -> mutateShardId(instance);
+            case 1 -> mutateCurrentNodeId(instance);
+            case 2 -> mutateState(instance);
+            default -> randomValueOtherThan(instance, this::createTestInstance);
+        };
+    }
+
+    private static ShardRouting mutateShardId(ShardRouting instance) {
+        var newShardId = switch (randomInt(2)) {
+            case 0 -> new ShardId(
+                randomValueOtherThan(instance.getIndexName(), ESTestCase::randomIdentifier),
+                instance.shardId().getIndex().getUUID(),
+                instance.id()
+            );
+            case 1 -> new ShardId(
+                instance.getIndexName(),
+                randomValueOtherThan(instance.shardId().getIndex().getUUID(), UUIDs::randomBase64UUID),
+                instance.id()
+            );
+            case 2 -> new ShardId(
+                instance.getIndexName(),
+                instance.shardId().getIndex().getUUID(),
+                randomValueOtherThan(instance.id(), () -> randomIntBetween(0, 99))
+            );
+            default -> throw new RuntimeException("unreachable");
+        };
+        return new ShardRouting(
+            newShardId,
+            instance.currentNodeId(),
+            instance.relocatingNodeId(),
+            instance.primary(),
+            instance.state(),
+            instance.recoverySource(),
+            instance.unassignedInfo(),
+            instance.relocationFailureInfo(),
+            instance.allocationId(),
+            instance.getExpectedShardSize()
+        );
+    }
+
+    private static ShardRouting mutateState(ShardRouting instance) {
+        var newState = randomValueOtherThan(instance.state(), () -> randomFrom(ShardRoutingState.values()));
+        return new ShardRouting(
+            instance.shardId(),
+            newState == ShardRoutingState.UNASSIGNED ? null : requireNonNullElseGet(instance.currentNodeId(), ESTestCase::randomIdentifier),
+            newState == ShardRoutingState.UNASSIGNED || newState == ShardRoutingState.STARTED
+                ? null
+                : requireNonNullElseGet(instance.relocatingNodeId(), ESTestCase::randomIdentifier),
+            instance.primary(),
+            newState,
+            newState != ShardRoutingState.UNASSIGNED && newState != ShardRoutingState.INITIALIZING
+                ? null
+                : requireNonNullElseGet(
+                    instance.recoverySource(),
+                    () -> TestShardRouting.buildRecoveryTarget(instance.primary(), newState)
+                ),
+            newState != ShardRoutingState.UNASSIGNED
+                ? instance.unassignedInfo()
```

Comment:
I am surprised we allow to keep unassigned info when shard is started or relocating

Reply:
Me too. Can we strengthen an assertion there?

Reply:
Will do in a followup